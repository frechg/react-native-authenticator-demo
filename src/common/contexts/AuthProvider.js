import React from 'react'
import * as SecureStore from 'expo-secure-store';
import { initialAuthState, authStateReducer } from '../authStateReducer';
import * as Api from '../../services/auth';

export const AuthContext = React.createContext();

export const AuthProvider = (props) => {
  const [authState, authDispatch] = React.useReducer(authStateReducer, initialAuthState);

  const getAuthState = async () => {
    try {
      const token = await SecureStore.getItemAsync('authToken');
      const username = await SecureStore.getItemAsync('username');
      authDispatch({type: 'RESTORE_TOKEN', token: token, username: username});
    } catch (error) {
      throw new Error(error);
    }
  };

  const signIn = async (data) => {
    await Api.signIn(data)
    .then(async (response) => {
      if (response.ok) {
        const jsonResponse = await response.json();
        await SecureStore.setItemAsync('authToken', jsonResponse.token);
        await SecureStore.setItemAsync('username', jsonResponse.username);
        authDispatch({type: 'SIGN_IN', token: jsonResponse.token, username: jsonResponse.username});
      } else {
        throw new Error('Email or password are incorrect.');
      }
    })
  };

  const signUp = async (data) => {
    await Api.signUp(data)
    .then(async (response) => {
      if (response.ok) {
        const jsonResponse = await response.json();
        await SecureStore.setItemAsync('authToken', jsonResponse.token);
        await SecureStore.setItemAsync('username', jsonResponse.username);
        authDispatch({type: 'SIGN_IN', token: jsonResponse.token, username: jsonResponse.username});
      } else {
        const jsonResponse = await response.json();
        throw new Error(jsonResponse.errors[0])
      }
    });
  };

  const signOut = async () => {
    try {
      await SecureStore.deleteItemAsync('authToken');
      await SecureStore.deleteItemAsync('username');
      authDispatch({type: 'SIGN_OUT'});
    } catch (error) {
      throw new Error(error);
    }
  };

  const contextValue = React.useMemo(() => {
    return {authState, getAuthState, signUp, signIn, signOut};
  }, [authState]);

  return (
    <AuthContext.Provider value={contextValue}>
      {props.children}
    </AuthContext.Provider>
  );
};