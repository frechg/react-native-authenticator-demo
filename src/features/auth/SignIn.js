import React from 'react';
import {
    Text,
    View,
    Button,
    TouchableOpacity,
    TouchableWithoutFeedback,
    Keyboard,
    KeyboardAvoidingView,
    Platform
} from 'react-native';
import { BasicTextInput } from '../../common/BasicTextInput';
import { BasicSubmitButton } from '../../common/BasicSubmitButton';
import { styles, formStyles } from '../../common/styles';
import { AuthContext } from '../../common/contexts/AuthProvider';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';

export const SignIn = ({ navigation }) => {
  const { signIn } = React.useContext(AuthContext);

  const validation = Yup.object({
    email: Yup.string().email('Invalid email').required('Required'),
    password: Yup.string().required('Required'),
  });

  const handleSubmit = async (values) => {
    await signIn(values)
  }

  return (
    <KeyboardAvoidingView 
      behavior={Platform.OS === "ios" ? "padding" : "height"}
      style={styles.container}
    >
      <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
        <View style={formStyles.formWrapper}>
          <Formik
            initialValues={{email: '', password: ''}}
            validationSchema={validation}
            onSubmit={async (values, {setSubmitting, setStatus}) => {
              await handleSubmit(values).catch((error) => {
                setStatus(error.message)
                setSubmitting(false);
              });
              setSubmitting(false);
            }}
          >
            {({status}) => (
              <View>
                { status && <Text style={formStyles.formError}>{status}</Text> }
                <Field
                  component={BasicTextInput}
                  name="email"
                  placeholder="Email"
                  keyboardType='email-address'
                />
                <Field
                  component={BasicTextInput}
                  name="password"
                  placeholder="Password"
                  secureTextEntry={true}
                />
                <BasicSubmitButton title='Sign In' />
              </View>
            )}
          </Formik>
          <TouchableOpacity style={styles.buttonSecondary} title='Create Account' onPress={() => navigation.navigate('Create Account')}>
            <Text style={styles.buttonSecondaryText}>Create Account</Text>
          </TouchableOpacity>
          <View style={{collr: 'rgb(100,100,100', marginTop: 10, flexDirection: 'row', justifyContent: 'center'}}>
            <Button
              title='Forgot Password'
              color='rgb(120,120,120)'
              onPress={() => navigation.navigate('Forgot Password')}
            />
          </View>
        </View>
      </TouchableWithoutFeedback>
    </KeyboardAvoidingView>
  );
}