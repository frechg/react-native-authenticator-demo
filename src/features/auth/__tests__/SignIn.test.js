import * as React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react-native';

import { SignIn } from '../SignIn';
import { AuthContext } from '../../../common/contexts/AuthProvider';

describe('<SignIn />', () => {
  test('Renders SignIn screen', () => {
    const signIn = jest.fn();    
    render(
      <AuthContext.Provider value={{signIn}}>
        <SignIn/>
      </AuthContext.Provider>
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('User submits form with valid authorization data', () => {
    const signIn = jest.fn();
    const { getByText, getByPlaceholderText } = render(
      <AuthContext.Provider value={{signIn}}>
        <SignIn/>
      </AuthContext.Provider>
    );

    // Setup form data
    const email = 'person@example.com';
    const password = 'testingpass';
    const formData = {email: email, password: password};

    // Get the form elements
    const emailInput = getByPlaceholderText('Your email');
    const passwordInput = getByPlaceholderText('Your password');
    const submitButton = getByText('Sign In');

    // Enter data and submit form
    fireEvent.changeText(emailInput, email);
    fireEvent.changeText(passwordInput, password);
    fireEvent.press(submitButton);

    expect(signIn).toHaveBeenCalledWith(formData);
  });

  test('User submits form without email', () => {
    const signIn = jest.fn();
    const { getByText, getByPlaceholderText } = render(
      <AuthContext.Provider value={{signIn}}>
        <SignIn/>
      </AuthContext.Provider>
    );
  
    fireEvent.changeText(getByPlaceholderText('Your password'), 'testingpass');
    fireEvent.press(getByText('Sign In'));
  
    expect(getByText('Email is required')).toBeDefined();
  });
  
  test('User submits form without password', () => {
    const signIn = jest.fn();
    const { getByText, getByPlaceholderText } = render(
      <AuthContext.Provider value={{signIn}}>
        <SignIn/>
      </AuthContext.Provider>
    );

    fireEvent.changeText(getByPlaceholderText('Your email'), 'person@example.com');
    fireEvent.press(getByText('Sign In'));

    expect(getByText('Password is required')).toBeDefined();
  });

  test('User sees error when sign in request fails', async () => {
    // Mock signIn to reject with an error
    const errorMessage = 'Sign in failed';
    const signIn = jest
      .fn()
      .mockRejectedValue(new Error(errorMessage));
    
    const { getByText, getByPlaceholderText } = render(
      <AuthContext.Provider value={{signIn}}>
        <SignIn/>
      </AuthContext.Provider>
    );

    await waitFor(() => {
      // Enter data and submit form
      fireEvent.changeText(getByPlaceholderText('Your email'), 'person@example.com');
      fireEvent.changeText(getByPlaceholderText('Your password'), 'testingpass');
      fireEvent.press(getByText('Sign In'));
      
      expect(getByText(errorMessage)).toBeDefined();
    });
  });
});