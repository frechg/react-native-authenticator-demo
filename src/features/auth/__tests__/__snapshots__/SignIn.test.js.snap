// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn /> Renders SignIn screen 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flex": 1,
        "justifyContent": "center",
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "padding": 32,
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "marginBottom": 24,
            "width": "100%",
          }
        }
      >
        <TextInput
          keyboardType="email-address"
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Email"
          style={
            Object {
              "borderBottomWidth": 1,
              "borderColor": "rgb(140,140,140)",
              "borderRadius": 0,
              "color": "rgb(60,60,60)",
              "fontSize": 18,
              "paddingBottom": 12,
              "paddingRight": 12,
              "paddingTop": 12,
              "width": "100%",
            }
          }
          value=""
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "marginBottom": 24,
            "width": "100%",
          }
        }
      >
        <TextInput
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Password"
          secureTextEntry={true}
          style={
            Object {
              "borderBottomWidth": 1,
              "borderColor": "rgb(140,140,140)",
              "borderRadius": 0,
              "color": "rgb(60,60,60)",
              "fontSize": 18,
              "paddingBottom": 12,
              "paddingRight": 12,
              "paddingTop": 12,
              "width": "100%",
            }
          }
          value=""
        />
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgb(40,40,40)",
            "borderRadius": 6,
            "opacity": 1,
            "padding": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(255,255,255)",
              "fontSize": 18,
            }
          }
        >
          Sign In
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgb(220,220,220)",
          "borderRadius": 6,
          "marginTop": 12,
          "opacity": 1,
          "padding": 12,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(60,60,60)",
            "fontSize": 18,
          }
        }
      >
        Create Account
      </Text>
    </View>
    <View
      style={
        Object {
          "collr": "rgb(100,100,100",
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": 10,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "rgb(120,120,120)",
                },
              ]
            }
          >
            Forgot Password
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
